lib/helpers.js

```diff
@@ -71,48 +71,61 @@ function fixImportedSourceMap(file, source, options, callback) {
   // remove source map comment from source
   file.contents = new Buffer(source.content, 'utf8');
 
-  async.forEachOf(source.map.sources, function(sourcePath, i, cb) {
-    if (matchesUrlRegex(sourcePath)) {
-      source.map.sourcesContent[i] = source.map.sourcesContent[i] || null;
+  async.eachOf(source.map.sources, normalizeSourcesAndContent, callback);
+
+  function assignSourcesContent(sourceContent, idx) {
+    source.map.sourcesContent[idx] = sourceContent;
+  }
+
+  function normalizeSourcesAndContent(sourcePath, idx, cb) {
+    var sourceRoot = source.map.sourceRoot || '';
+    var sourceContent = source.map.sourcesContent[idx] || null;
+
+    if (isRemoteSource(sourcePath)) {
+      assignSourcesContent(sourceContent, idx);
       return cb();
     }
-    var absPath = path.resolve(source.path, sourcePath);
-    source.map.sources[i] = unixStylePath(path.relative(file.base, absPath));
-    if (!source.map.sourcesContent[i]) {
-      var sourceContent = null;
-      if (source.map.sourceRoot) {
-        if (matchesUrlRegex(source.map.sourceRoot)) {
-          source.map.sourcesContent[i] = null;
-          return cb();
-        }
-        absPath = path.resolve(source.path, source.map.sourceRoot, sourcePath);
-      }
-      if (absPath === file.path) {
-        // if current file: use content
-        sourceContent = source.content;
-      } else {
-        // else load content from file async
-        // if (options.debug) {
-        //  console.log(PLUGIN_NAME + '-add: No source content for "' + source + '". Loading from file.');
-        // }
-        return fs.readFile(absPath, 'utf8', onRead);  
-      }
-      source.map.sourcesContent[i] = sourceContent;
+
+    if (source.map.sourcesContent[idx]) {
+      return cb();
     }
+
+    if (sourceRoot && isRemoteSource(sourceRoot)) {
+      assignSourcesContent(sourceContent, idx);
+      return cb();
+    }
+
+    var basePath = path.resolve(file.base, sourceRoot);
```

Comment:
@phated Afraid I can't really help much with that, also just refactored existing code and fixed the tests, they did look pretty thorough and added a few myself so hopefully it should be ok.